### ISSUE TITLE
API documentation revamp

### ISSUE THREAD

**Issue by hugolpz** [kanasimi/wikiapi#30]
### API actionables
- [x] Identify and list the core steps or js methods ([here](https://github.com/kanasimi/wikiapi/blob/master/wikiapi.js#L946-L988))
- [x] Demo initial documentation via D3js documentation style
- [x] Explore documentation generators
- [x] JSdoc3 implementation.
- [x] Docdash theme installation
- [x] Improve in-code comments accordingly to JSdoc3 syntax
- [x] Set and improve github actions

**Comment by hugolpz** [comment]
See best practice [d3-array](https://github.com/d3/d3-array/blob/v2.12.1/README.md)

<a name="max" href="#max">#</a> d3.<b>max</b>(<i>iterable</i>[, <i>accessor</i>]) · [Source](https://github.com/d3/d3-array/blob/master/src/max.js).
Returns .... If ..., returns.... An optional *option_name* does ....
See also [related_stuff](#related_stuff).

**Comment by hugolpz** [comment]
I'am starting work here. Help welcome : 

<a name="site_name" href="#site_name">#</a> .<b>site_name</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="login" href="#login">#</a> .<b>login</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="query" href="#query">#</a> .<b>query</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="page" href="#page">#</a> .<b>page</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="tracking_revisions" href="#tracking_revisions">#</a> .<b>tracking_revisions</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="edit_page" href="#edit_page">#</a> .<b>edit_page</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="edit" href="#edit">#</a> .<b>edit</b>(<i>p1</i>[,<i>options</i>]content, options) {}: {definition}
<a name="move_to" href="#move_to">#</a> .<b>move_to</b>(<i>p1</i>[,<i>options</i>]): move page to new title address.
* `new`: string.
* `options`: object. See …

<a name="move_page" href="#move_page">#</a> .<b>move_page</b>(<i>curr</i>, <i>new</i>[,<i>{options}</i>]): moves target page to new title attress.
* `curr`: string.
* `new`: string.
* `options`: object. See …

<a name="purge" href="#purge">#</a> .<b>purge</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="listen" href="#listen">#</a> .<b>listen</b>(<i>p1</i>[,<i>options</i>]listener, options) {}: {definition}
<a name="category_tree" href="#category_tree">#</a> .<b>category_tree</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="search" href="#search">#</a> .<b>search</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="redirects_root" href="#redirects_root">#</a> .<b>redirects_root</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="redirects_here" href="#redirects_here">#</a> .<b>redirects_here</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="register_redirects" href="#register_redirects">#</a> .<b>register_redirects</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="upload" href="#upload">#</a> .<b>upload</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="get_featured_content" href="#get_featured_content">#</a> .<b>get_featured_content</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="for_each_page" href="#for_each_page">#</a> .<b>for_each_page</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="for_each" href="#for_each">#</a> .<b>for_each</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="data" href="#data">#</a> .<b>data</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="convert_Chinese" href="#convert_Chinese">#</a> .<b>convert_Chinese</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="run_SQL" href="#run_SQL">#</a> .<b>run_SQL</b>(<i>p1</i>[,<i>options</i>]): {definition}
<a name="setup_layout_elements" href="#setup_layout_elements">#</a> .<b>setup_layout_elements</b>(<i>p1</i>[,<i>options</i>]): {definition}

**Comment by hugolpz** [comment]
6. `.upload()` needs review on the options side. 

```js
{ file_path: '/local/file/path', comment: '',
		description: '', date: new Date() || '2021-01-01', source_url: 'https://github.com/kanasimi/wikiapi', author: '[[User:user]]', permission: '{{cc-by-sa-2.5}}', other_versions: '', other_fields: '',
		license: ['{{cc-by-sa-2.5}}'], categories: ['[[Category:test images]]'],
		// To overwrite existing files
		ignorewarnings: 1
}
```
7. Some wikidata specific methods where not in [here](https://github.com/kanasimi/wikiapi/blob/master/wikiapi.js#L946-L988) : entity, modify, others ?

**Comment by kanasimi** [comment]
Maybe we can use https://github.com/jsdoc/jsdoc ?

**Comment by hugolpz** [comment]
**TL;DR:** expand readme above will be faster.

**Long version :** I'am discovering JSdoc3. On the long run it's a very good idea. But as of now I lack visibility. Your wikiapi.js code currently has (i think) a mix of comment syntaxes. I have no idea how it will play out with installing JSdoc. If you are familiar with JSdoc and can smoothly go ahead, then the sooner the better.

If not, I would suggest to rather complete and improve the markdown draft above. Within 1~2h, WikiAPI will have a clean doc for +90% of the code and methods. This will be concise and elegant. (JSdoc rendering are not concise).
(Also I will be less active in coming week, must catch up upon IRL todo list.)

Later on, with few regex, this markdown can be converted back into JSdoc3. When ready : wikiapi.js comments cleaned up, you/we properly learnt JSdoc3 syntaxes, which template to use.

```markdown
<a name="move_to" href="#move_to">#</a> .<b>move_to</b>(<i>move_to_title</i>[,<i>options</i>]): move page to new title address.
* `move_to_title`: string.
* `options`: object.
  * `reason`: string.
  * `noredirect`: boolean
  * `movetalk`: boolean.
```
...into...

```javascript
/**
 * Moves page to new title address.
 * @param {String} move_to_title - target page's title.
 * @param {Object} options - options.
 * @param {String} options.reason - reason for the move.
 * @param {Boolean} options.noredirect - do not leave a redirect.
 * @param {Boolean} options.movetalk - move the talkpage as well.
 * @returns {Object} - returns (does not exist on .move_to but I put it here for demo)
 */
function move_to (move_to_title,options) { 
    // do stuff here
}
```

**Comment by kanasimi** [comment]
try https://kanasimi.github.io/wikiapi/

**Comment by hugolpz** [comment]
Nice :sunglasses

**Comment by hugolpz** [comment]
Saw the changes. Starts to be awesome !

**Comment by kanasimi** [comment]
Substantial changes of documents have ended now.

**Comment by hugolpz** [comment]
Good work done. Close. Thank @kanasimi !
